### PR TITLE
Mobt 157 nbhood refactor consolidate unit tests rebased

### PR DESCRIPTION
**Pull request by fionaRust**
partially addresses https://github.com/metoppv/mo-blue-team/issues/157

Moves unit tests around and simplifies some interfaces.

Note: reviewer should look at new unit tests in NeighbourhoodProcessing and check they are happy with coverage, then look at unit tests removed.

Most of square ones have been repurposed for new NeighbourhoodProcessing unit tests. Circular ones had lots of edge cases for circular neighbourhood behaviour, but these are really demonstrating underlying scipy functionality not anything else, so removed. Base neighbourhood processing unit tests didn't really add anything new, so got rid of those too.

Percentiles neighbourhood tests just moved with little change, so should separate out into separate PR. Could also move some of my other tidying things up and moving things around commits into another PR. - THis part is now moved into https://github.com/metoppv/improver/pull/1665 which has been done first.

### CODE REVIEW

**Review by MoseleyS — CHANGES_REQUESTED** [review]
I think this is basically fine. Just a couple of suggestions.

File: improver_tests/nbhood/square_kernel/test_SquareNeighbourhood.py [hunk 1]
```diff
@@ -1,418 +0,0 @@
...
-                [1.0, 0.88888889, 0.88888889, 0.88888889, 1.0],
-                [1.0, 0.88888889, 0.88888889, 0.88888889, 1.0],
-                [1.0, 1.0, 1.0, 1.0, 1.0],
-            ]
-        )
-        result = NeighbourhoodProcessing("square", self.RADIUS).process(self.cube)
-        self.assertIsInstance(result, Cube)
-        self.assertArrayAlmostEqual(result.data, expected_array)
-
-    def test_negative_strides_re_mask_true(self):
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply improver_tests/nbhood/square_kernel/test_SquareNeighbourhood.py:83]: I think this does actually still work, do you want me to put a test back in to show it does?

File: improver_tests/nbhood/nbhood/test_NeighbourhoodProcessing.py [hunk 2]
```diff
@@ -34,88 +34,338 @@
...
+    def test_masked_array_re_mask_true_square(self):
+        """Test the _calculate_neighbourhood method when masked data is
+        passed in and re-masking is applied."""
+
+        input_data = np.ma.masked_where(self.mask == 0, self.data_for_masked_tests)
+        plugin = NeighbourhoodProcessing("square", self.RADIUS)
+        plugin.nb_size = self.nbhood_size
+        result = plugin._calculate_neighbourhood(input_data)
+        self.assertArrayAlmostEqual(result.data, self.expected_array)
+        self.assertArrayAlmostEqual(result.mask, self.expected_mask)
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply improver_tests/nbhood/nbhood/test_NeighbourhoodProcessing.py:222]: done

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks @fionaRust 👍 

I've added some very minor comments.

File: improver_tests/nbhood/nbhood/test_NeighbourhoodProcessing.py [hunk 3]
```diff
@@ -34,88 +34,338 @@
...
+        plugin.kernel = self.weighted_circular_kernel
+        result = plugin._calculate_neighbourhood(self.data)
+        self.assertArrayAlmostEqual(result.data, expected_array)
+
+    def test_basic_square_sum(self):
+        """Test the _calculate_neighbourhood method calculating a sum in
+        a square neighbourhood."""
+        expected_array = np.array(
+            [
+                [4.0, 6.0, 6.0, 6.0, 4.0],
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply improver_tests/nbhood/nbhood/test_NeighbourhoodProcessing.py:184]: Yeah, you're right. I have discussed this on the main ticket. https://github.com/metoppv/mo-blue-team/issues/157#issuecomment-1023210150 I think this is something we should come back to if we want to use circular more, but speed is still a limiting factor atom

File: improver/nbhood/square_kernel.py [hunk 4]
```diff
@@ -111,7 +116,9 @@ def __init__(
         self.sum_only = sum_only
         self.re_mask = re_mask
 
-    def _calculate_neighbourhood(self, data: ndarray, mask: ndarray) -> ndarray:
+    def _calculate_neighbourhood(
+        self, data: ndarray, mask: ndarray = None
+    ) -> Union[ndarray, np.ma.MaskedArray]:
```
Comment (gavinevans) [inline improver/nbhood/square_kernel.py:121]: As mentioned as something that could be considered for a separate PR, I think that potentially the following code block in this method can be simplified as shown below. This is because the "external mask" provided using the argument to the method is already incorporated into the full mask (combining the externally provided mask with a mask from the array) on line 148 and 151.

```
        area_mask_dtype = np.int64
        if mask is None:
            area_mask = np.ones(data.shape, dtype=area_mask_dtype)
        else:
            area_mask = np.array(mask, dtype=area_mask_dtype, copy=False)
```

to

```
area_mask = np.ones(data.shape, dtype=np.int64)
```

File: improver/nbhood/square_kernel.py [hunk 5]
```diff
@@ -111,7 +116,9 @@ def __init__(
         self.sum_only = sum_only
         self.re_mask = re_mask
 
-    def _calculate_neighbourhood(self, data: ndarray, mask: ndarray) -> ndarray:
+    def _calculate_neighbourhood(
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply improver/nbhood/square_kernel.py:119]: I'll do this in a follow up PR

File: improver_tests/nbhood/nbhood/test_NeighbourhoodProcessing.py [hunk 6]
```diff
@@ -124,11 +374,25 @@ def test_square_neighbourhood(self):
...
         radii = 2000
-        result = NBHood(neighbourhood_method, radii)(self.cube)
+        result = NeighbourhoodProcessing(neighbourhood_method, radii)(self.cube)
         self.assertIsInstance(result, Cube)
         self.assertArrayAlmostEqual(result.data, expected)
+        self.assertTupleEqual(result.cell_methods, self.cube.cell_methods)
+        self.assertDictEqual(result.attributes, self.cube.attributes)
+
+    def test_cube_metadata(self):
+        """Test that the square neighbourhood processing is successful."""
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply improver_tests/nbhood/nbhood/test_NeighbourhoodProcessing.py:381]: I'll do this in a follow up PR

File: improver/nbhood/square_kernel.py (hunk 4 above)
Comment (fionaRust) [reply improver/nbhood/square_kernel.py:121]: I'll do this in a follow up PR

File: improver_tests/nbhood/square_kernel/test_SquareNeighbourhood.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver_tests/nbhood/square_kernel/test_SquareNeighbourhood.py:83]: I think no. It doesn't fit with the IMPROVER philosophy to support negatively-striding coordinates, so let's drop this.

**Review by MoseleyS — APPROVED** [review]

**Review by gavinevans — APPROVED** [review]
